### ISSUE TITLE
Linux, Makefile, cli-dependencies target references msbuild

### ISSUE THREAD

**Issue by anvilvapre** [OpenRA/OpenRA#16615]
Older versions of mono-devel do not provide msbuild. 
The Makefile, target core, checks to see if msbuild is present, warns if not. 
Yet msbuild is already referenced earlier in cli-dependencies. 
Solution: adjust Makefile to do check earlier. 
Prio: minimal, low, development env only
Reporting it because it may keep new contributers from getting up and running easily.

**Comment by pchote** [comment]
Removing the `msbuild` dep is a wontfix, but we should fail earlier if it is missing.

**Comment by Mailaender** [comment]
https://github.com/mono/linux-packaging-msbuild but you probably won't be able to package it as it installs dotnet during the build scripts and recovers several nuget packages.

**Comment by pchote** [comment]
See also https://github.com/OpenRA/OpenRA/issues/17508#issuecomment-568939206.

**Comment by Mailaender** [comment]
My current workaround due to https://github.com/mono/monodevelop-flatpak/issues/9 is

```
wget flatpak install --user --from https://download.mono-project.com/repo/monodevelop.flatpakref
```

Remove the line to the GNOME SDK as it is 404.

```
flatpak install --user --from *.flatpakref
```

You get a broken and unmaintained version of MonoDevelop, but it compiles OpenRA, but it will then crash the IDE after compilation errors somehow. Don't try to run the game as the sandbox doesn't have enough rights to get accelerated graphics.

**Comment by anvilvapre** [comment]
it's not a problem for developers to install msbuild themselves is it, with some instructions.
the issue was more about a build failing because of unchecked missing dependencies.

**Comment by Mailaender** [comment]
Actually getting msbuild on @openSUSE is currently a huge problem for developers that I haven't solved for myself.

**Comment by anvilvapre** [comment]
i see. using Ubuntu.

**Comment by Mailaender** [comment]
Made an upstream ticket https://github.com/mono/linux-packaging-msbuild/issues/10

**Comment by AptiviCEO** [comment]
This repo should build MSBuild offline: https://github.com/EoflaOE/linux-remote-packaging-msbuild
Also, look at https://github.com/EoflaOE/linux-remote-packaging-msbuild/blob/master/documentation/building-msbuild-offline.md for instructions on how to build it **offline**.

**Comment by Mailaender** [comment]
Thanks! Can you re-upload the dependencies https://github.com/EoflaOE/linux-remote-packaging-msbuild/releases/download/bootstrap%2F16.5%2Bxamarinxplat.2020.02.20.11.54/Pkgs-CoreEssentials.tar.gz with proper filenames / directories or rather ones that don't need to be escaped in an rpm .spec file?

**Comment by Mailaender** [comment]
Can't get it to work though. A whole folder `.dotnet` seems to be missing:
```
[    7s] + cd linux-remote-packaging-msbuild-bootstrap-16.5-xamarinxplat.2020.02.20.11.54
[    7s] + ./build.sh
[    7s] Bootstrapping '/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-bootstrap-16.5-xamarinxplat.2020.02.20.11.54/.dotnet/../required/.dotnet/dotnet-install.sh'
[    7s] cp: cannot stat '/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-bootstrap-16.5-xamarinxplat.2020.02.20.11.54/.dotnet/../required/.dotnet/dotnet-install.sh': No such file or directory
[    7s] Failed to bootstrap dotnet install script (exit code '1').
```

**Comment by AptiviCEO** [comment]
I have fixed the .dotnet folder being missing. Can you retry building, but this time, clone the source code from master?

About the reupload, I will do it once I get myself a CentOS VM to be RPM .spec friendly.

**Comment by Mailaender** [comment]
Something is wrong with the directory structure. It is digging `..` way too deep:

```
[    6s] + cd linux-remote-packaging-msbuild-master
[    6s] + ./build.sh
[    6s] Bootstrapping '/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/.dotnet/../required/.dotnet/dotnet-install.sh'
[    6s] dotnet-install: Verifying file: /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/../../../../required/dotnet-sdk-2.1.7-linux-x64.tar.gz
[    6s] cp: cannot stat '/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/../../../../required/dotnet-sdk-2.1.7-linux-x64.tar.gz': No such file or directory
[    6s] dotnet-install: Cannot download: /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/../../../../required/dotnet-sdk-2.1.7-linux-x64.tar.gz
[    6s] dotnet_install: Error: Could not find/download: `.NET Core SDK` with version = 2.1.7
[    6s] dotnet_install: Error: Refer to: https://aka.ms/dotnet-os-lifecycle for information on .NET Core support
[    6s] Failed to install dotnet SDK from public location (exit code '1').
[    6s] /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/eng/common/tools.sh: line 389: /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/.dotnet/dotnet: No such file or directory
[    6s] Build failed (exit code '127').
```
Are there any instructions on how to use `build.sh`?

**Comment by AptiviCEO** [comment]
Not yet for CentOS. I will try to make the script avoid using `../../../../` in master if building in Red Hat and its derivatives by checking to see if building on Debian and its derivatives once I get my CentOS ready. Just edit the `required/.dotnet/dotnet-install.sh` to remove the `/../../../..` from azure_feed and uncached_feed in lines 895 and 896.

The reason that it failed is that because in `msbuild.spec`, there is no `HOME=` prefix before build scripts in lines 76 and 80, while on Debian's `rules`, there is the same prefix before build scripts in lines 14 and 17.

**Comment by AptiviCEO** [comment]
I have tested it on my clean CentOS dev virtual machine, and it worked without giving the error you've mentioned (`rpmbuild --target noarch -bb msbuild.spec`) until it stopped on `assembly not found` messages. I thought you were using CentOS, but you're using OpenSUSE. I will give that a try and keep you up to date.

**Comment by Mailaender** [comment]
Can you share you .spec file? I tried starting the scripts like the Debian rules does, but it doesn't work:

```
[    7s] + /usr/bin/rm -rf /home/abuild/rpmbuild/BUILDROOT/msbuild-16.5+xamarinxplat.2020.02.20.11.54-0.x86_64
[    7s] ++ dirname /home/abuild/rpmbuild/BUILDROOT/msbuild-16.5+xamarinxplat.2020.02.20.11.54-0.x86_64
[    7s] + /usr/bin/mkdir -p /home/abuild/rpmbuild/BUILDROOT
[    7s] + /usr/bin/mkdir /home/abuild/rpmbuild/BUILDROOT/msbuild-16.5+xamarinxplat.2020.02.20.11.54-0.x86_64
[    7s] + cd linux-remote-packaging-msbuild-master
[    7s] + mkdir -p /home/abuild/rpmbuild/BUILDROOT/msbuild-16.5+xamarinxplat.2020.02.20.11.54-0.x86_64/homedir/
[    7s] + HOME=/home/abuild/rpmbuild/BUILDROOT/msbuild-16.5+xamarinxplat.2020.02.20.11.54-0.x86_64/homedir/
[    7s] + ./eng/cibuild_bootstrapped_msbuild.sh --host_type mono --configuration Release --skip_tests /p:DisableNerdbankVersioning=true
[    7s] ** Bootstrapping MSBuild...
[    7s] Microsoft (R) Build Engine version 16.3.0-ci for Mono
[    7s] Copyright (C) Microsoft Corporation. All rights reserved.
[    7s] 
[    7s] Some command line switches were read from the auto-response file "MSBuild.rsp". To disable this file, use the "-noAutoResponse" switch.
[    7s] /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/eng/../artifacts/mono-msbuild/MSBuild.dll /fl /bl /clp:Summary /flp:v=diag /m:1 /p:MSBuildExtensionsPath=/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/eng/../artifacts/mono-msbuild/Extensions /p:MSBuildExtensionsPath32=/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/eng/../artifacts/mono-msbuild/Extensions /p:MSBuildExtensionsPath64=/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/eng/../artifacts/mono-msbuild/Extensions /p:DeterministicSourcePaths=false /p:MicrosoftNetCompilersVersion=3.5.0-beta4-20121-02 /p:NuGetBuildTasksVersion=5.5.0-rtm.6422 /verbosity:minimal mono/build/update_bundled_bits.proj
[    8s] /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/mono/build/update_bundled_bits.proj(2,5): error MSB4019: The imported project "/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/mono/build/DotnetBitsVersions.props" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
[    8s] 
[    8s] Build FAILED.
[    8s] 
[    8s] /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/mono/build/update_bundled_bits.proj(2,5): error MSB4019: The imported project "/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-master/mono/build/DotnetBitsVersions.props" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
[    8s]     0 Warning(s)
[    8s]     1 Error(s)
```

**Comment by Mailaender** [comment]
See https://build.opensuse.org/package/show/home:Mailaender:branches:Mono:Factory/msbuild for mine.

**Comment by AptiviCEO** [comment]
This is my spec file: https://pastebin.com/qUSs0LCM

From now on, remove the `HOME=` prefix before the script. I will fix the `dotnet-install.sh` script so when the build environment is OpenSUSE Build Service, the `azure_feed` and `uncached_feed` will no longer include the trailing `/../../../..` which makes the build fail with the output in [this comment](https://github.com/OpenRA/OpenRA/issues/16615#issuecomment-633284779)

**Comment by AptiviCEO** [comment]
Prepared patch for fixing the dotnet-install.sh path issue when building on OpenSUSE Build Service: https://github.com/EoflaOE/linux-remote-packaging-msbuild/blob/centos/osc-patches/fix-too-deep-directory-lookup.patch

**Comment by Mailaender** [comment]
That one does not apply to master. Do I have to use a specific branch or tag?

**Comment by AptiviCEO** [comment]
Just place the patch in the source files root directory (the same directory where the msbuild.spec and tar.gz archives exist), and add the new entry for patch to msbuild.spec. Let me know if it patched the dotnet-install.sh.

**Comment by Mailaender** [comment]
This finally forces me to fix
> error MSB4242: The SDK resolver "Microsoft.DotNet.MSBuildSdkResolver" failed to run.

which I now did in https://build.opensuse.org/request/show/809078.

**Comment by AptiviCEO** [comment]
Nice. In msbuild, have you managed to use the dotnet-install.sh which removes the `/../../../..`? Did you get the build to succeed with the patch?

**Comment by Mailaender** [comment]
I managed to succeed the build and apparently I didn't need the patch. This is built from your master https://build.opensuse.org/package/rdiff/home:Mailaender:branches:Mono:Factory/msbuild?opackage=msbuild&oproject=Mono%3AFactory&rev=5 I will try to rebase your patches on the latest official release now.

**Comment by AptiviCEO** [comment]
Nice. If you managed to rebase my patches on the official release with success, tell me.

**Comment by Mailaender** [comment]
How did you manage to dump the artifacts? Even after applying https://github.com/mono/linux-packaging-msbuild/pull/14 I am blocked by https://github.com/mono/linux-packaging-msbuild/issues/15.

**Comment by AptiviCEO** [comment]
Do you mean by "artifacts" the files that are stored in "required" folder and necessary dependencies that are stored in "Pkgs-Essentials.tar.gz" and "Pkgs-CoreEssentials.tar.gz"? If so, then, I have first waited for the build process to complete. Next, I have grabbed the links of the MSBuild bootstrapper, the dotnet-install.sh script, and the .NET Core runtime for **x64** Linux, and downloaded them to required. Then, I have grabbed all the required packages by copying the directory called debian/homedir/.nuget to required. Finally, I have tarred and gzipped both the required/.nuget and .packages directories.

About the error, are you running the 32-bit version of openSUSE or the 64-bit one? Sounds like that may be the reason, as the 64-bit version of .NET Core is used.

**Comment by Mailaender** [comment]
openSUSE deprecated it's 32 bit versions and .NET Core only supports x64 for openSUSE for me.

**Comment by Mailaender** [comment]
It looks like the dependencies for version 16.5-xamarinxplat.2020.02.20.11.54 like Microsoft.NET.Sdk version='3.1.200-preview.20113.2' are gone now and can't be restored.

> error MSB4236: Das angegebene SDK "Microsoft.DotNet.Arcade.Sdk" wurde nicht gefunden.

**Comment by AptiviCEO** [comment]
I've looked at https://dotnetfeed.blob.core.windows.net/dotnet-core/sleet.packageindex.json and saw that it's there. Is it from my offline MSBuild fork? I have looked in Pkgs-Essentials.tar.gz now, and I saw that it's found. The Microsoft.Net.Sdk one and the Microsoft.DotNet.Arcade.Sdk one.

**Comment by Mailaender** [comment]
Can you somehow submit your changes upstream? It seems to be a nightmare to update this.

**Comment by AptiviCEO** [comment]
I will try to do it. However, it takes me days because I have limited resources (electricity being off 2 hours and on 4 hours). It would have taken less than a day if the electricity issue didn't exist in my area. When it's ready, I will notify you.

**Comment by Mailaender** [comment]
I saw that you updated to version 16.6 recently. Do you have the `Pkgs.tar.gz` at hand as well?

**Comment by AptiviCEO** [comment]
I have updated it to 16.6 but didn't upload the required package tarballs. The upload speed is slow so it's going to take a while before they upload. Extract both tars to the root folder of the repository.

**Comment by Mailaender** [comment]
You could also provide a script on how to generate those tarballs. I can then upload them for you.

**Comment by AptiviCEO** [comment]
Thanks. I will try implementing a script that can generate those tarballs when I have free time. However, it needs to download packages and the required files to build it because I shipped those tarballs with not only the required nuget packages, but also the dotnet SDK and runtime, the dotnet-install.sh, and the msbuild.zip bootstrapper. It will not run during the build, though, so it can be used to pack such files.

I also release such tarballs in the Releases of the repository, so I will add you to it once I finish implementing the script so you can upload these files for me.

**Comment by AptiviCEO** [comment]
@Mailaender Would you be able to test the script for me? I have very limited time to work on the computer because of the internal restrictions, including the Internet full speed quota which is about 85 GB, me not being allowed to work on the PC more than 2/3 hours, and the electricity being off/on.

Before publishing those tarballs, verify that the generated `required` folder works when offline-compiling and that there are no missing NuGet packages. If any errors, tell me. The script will leave the `required` folder alone for compilation.

**Comment by Mailaender** [comment]
Sure, which script/command shall I run?

**Comment by AptiviCEO** [comment]
1. Check the updated script to fix missing packages issues.
2. The script name is generate_tarballs.sh

You can run this script without any arguments provided. It uses the Internet connection but will not be executed during the build process. Run it using `./generate_tarballs.sh`.

**Comment by Mailaender** [comment]
Uploaded to https://github.com/EoflaOE/linux-remote-packaging-msbuild/releases/tag/untagged-f69a60ee3df292fee54d

**Comment by Mailaender** [comment]
Doesn't build though:
```
[   26s] Build FAILED.
[   26s] 
[   26s] CSC : warning CS2008: No source files specified. [/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-bootstrap-16.6-xamarinxplat.2020.04.29.14.43/mono/build/sdks_and_nugets/nuget-build-tasks/nuget-build-tasks.csproj]
[   26s] dotnet_install : error : curl (recommended) or wget are required to download dotnet. Install missing prerequisite to proceed. [/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-bootstrap-16.6-xamarinxplat.2020.04.29.14.43/mono/build/sdks_and_nugets/update_sdks_and_nugets.proj]
[   26s] /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-bootstrap-16.6-xamarinxplat.2020.04.29.14.43/mono/build/sdks_and_nugets/update_sdks_and_nugets.proj(64,9): error MSB3073: The command "/home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-bootstrap-16.6-xamarinxplat.2020.04.29.14.43/mono/build/sdks_and_nugets/../get_sdk_files.sh /home/abuild/rpmbuild/BUILD/linux-remote-packaging-msbuild-bootstrap-16.6-xamarinxplat.2020.04.29.14.43/mono/build/../../mono/dotnet-overlay/msbuild-bin" exited with code 1.
```

**Comment by AptiviCEO** [comment]
OK. I will fix it soon. You'll be notified if the change is made.

**Comment by AptiviCEO** [comment]
I have updated the repository to fix the build failure if the target has no curl/wget. You may need to reupload the first CoreEssentials.tar.xz, but no need to reupload the second.

Tell me if there are other build failures.

**Comment by Mailaender** [comment]
Updated the first binary and prepared a small patch https://github.com/EoflaOE/linux-remote-packaging-msbuild/pull/2 otherwise the build needs `wget` or `curl` although that is never used. See https://build.opensuse.org/request/show/825133 for an integration test which now runs fine.

**Comment by AptiviCEO** [comment]
Thanks for the info! I will update my PPA soon with the updated version.

**Comment by anvilvapre** [comment]
Can this be closed?

**Comment by pchote** [comment]
The `cli-dependencies` target was removed and we now only have two places where we call msbuild. One of them checks and gives a warning but the other doesn't. Since mono is no longer the default, and the user must explicitly opt in to using it, I think it would be reasonable to remove the warning completely instead of adding it to the other place.

**Comment by anvilvapre** [comment]
The warning also doesn't harm. Can be removed one day.
